Feat  fsa2021 124 add play plus test

FlashGooden opened the pull request:
## Description
Added Play function With tests...... I know this is a big PR.... I'm sorry. 
I tried to curate all the Commit notes so you can see what is going on. 

## Spec

See Issue: [ISSUE_NUMBER](https://sparkbox.atlassian.net/browse/FSA2021-124)

## Validation

- [x] This PR has code changes, and our linters still pass.
- [x] This PR has new code, so new tests were added or updated, and they pass.

### To Validate

1. Make sure all PR Checks have passed.
2. Pull down all related branches.
3. Navigate to... `feat--FSA2021-124-add-play-plus-test`
4. run npm test
5. run npm start

FlashGooden committed:
chore: installing mp3 codecs
- we need these two libraries to play mp3 files in discord

FlashGooden committed:
feat: First edition of play (Deprecated and renewed in further commit)
- our mp3Path variable is our local file location for our sound
- our streampath var is a link to an mp3
- Right now play function will:
1.check if a member s in a voice channel
2.they they are then the bot will join the channel
3.the bot will play the file handed to it and console.log that it is finished playing or say you need to be in a voice channel

FlashGooden committed:
refactor: abstract play function
-abstracted play function by having a playMp3 function
-playMp3 will take in a room connection object provided by discord
- the roomConnection object passed in will play an mp3 URL that it is passed
- I created an abstraction function for checking if a mbmer is in a voice channel
- I created an abstraction croomConnection that lets the bot join the voice channel
-(line 18) play function now uses those abstracted functions

FlashGooden committed:
refactor: Refactor index.js
- combined my procress.env variables
- truncated server messages
- passed in argument to discord bot

FlashGooden committed:
chore: xhr2 package
- Im going to have to use this package to assist with pulling infomration from Firebase

FlashGooden committed:
chore: adding config for firebase

FlashGooden committed:
Refactor & Feat
I was looking to get a feature working so I imported the files needed for firebase into this commit. This made me have to alter my playMp3 function to incorporate the firebase api
- (line 1 -2 ) we import the firebase api
- we import xhr2 needed for our firebase link requests
- firebaseConfig is all our info needed to identify the storage cluster
- firebaseApp initializes firebase
- (line 21-24) are references to all our firebase sounds
- (26-33) is how we transform our references into URLS from firebase
- refactor playMp3 :
1. using the new arg parameter passed in we compare it against the sound references then plays with the corresponding URL or return null

FlashGooden committed:
refactor: updated messages from Help and List js

FlashGooden committed:
chore: install jest for testing

FlashGooden committed:
Feature: Playing Stream + Firebase sourcing + Play.js Tests
This  is a big PR so excuse the mess. This is a combination of FSA2021-124 and FSA2021-128
The main things done in this PR a:
- Getting our Play.js to play a sound file
- Sourcing the sound files from Firebase
- Adding tests

FlashGooden committed:
chore: ESLint update config

FlashGooden committed:
test: Play.js tests
- added a test to give an error if play function fails to be called without being in a voice channel
- added a test that checks to see if a user is in a room where the mp3 should be played.

FlashGooden committed:
chore: babel config

FlashGooden committed:
chore: PR template added

catheraaine commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,18 @@
+## Description
```
Excellent!

catheraaine commented on Commands/Play.spec.js:
```diff
@@ -0,0 +1,30 @@
+import play from "./Play";
+
+describe("Tests the play function", () => {
```
You can think of your test descriptions as a sentence without the keywords. For example, as currenntly written, this would say: 

"Tests the play function will error if you are not in a voice channel" 
"Test the play function will join channel to play mp3 if command is available and user is in room" 

Which sounds better if you marked it up like this

```js
describe("The Play Function", () => {
  it("throws an error if you are not in a voice channel", async () => {
  it("joins a channel to play mp3 if the command is available and user is in room", async () => {

catheraaine replied:
As in, "The Play Function throws an error if you are not in a voice channel." sounds like a sentence.

FlashGooden replied:
Noted, I will change this. Thank you.

catheraaine commented on babel.config.cjs:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  presets: [['@babel/preset-env', {targets: {node: 'current'}}]],
+};
```
you could expand this out so it's easier to read, but you don't have to. I think you have an extra set of `[]`

```js
module.exports = {
  presets: [
    [
    '@babel/preset-env',
    {
      targets: {
        node: 'current'
      }
    }
   ]
  ],
};

catheraaine commented on index.js:
```diff
@@ -4,14 +4,13 @@ import play from './Commands/Play.js'
 
-const { BOTPREFIX } = process.env;
-const { TOKEN } = process.env;
+const { BOTPREFIX, TOKEN} = process.env;
```
nice update. add a space after token. Do you have a linter set up in your code editor?

FlashGooden replied:
I do, I'm surprised this didn't error, my rules are imported from the Discord docs, I'll check.

catheraaine commented on package.json:
```diff
@@ -6,7 +6,7 @@
   "scripts": {
     "start": "node index.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest --verbose"
```
nice nice

catheraaine replied:
you may want to add a lint task, too

catheraaine commented on package.json:
```diff
@@ -24,10 +24,19 @@
+    "@discordjs/opus": "^0.5.0",
     "discord.js": "^12.5.3",
-    "dotenv": "^8.2.0"
+    "dotenv": "^8.2.0",
```
is this one a dep or a dev dependency

FlashGooden replied:
Honestly I'm not sure. Dotenv is needed so I can grab variables located in my env file..... I googled it and some say leave it as a regular dependency. The Docs install it as a regular dependency as well. I'm going to lean more on regular dependency on this one because my app won't function without calling this library.

catheraaine replied:
> my app won't function without calling this library

Yep that's the definition of a dependency! 

A dev dependency is something like ESLint or typescript types files - things that your app doesn't need after it compiles.

catheraaine reviewed (commented):
Looking good so far, @Gamerman2001

jordanleven commented on babel.config.cjs:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  presets: [['@babel/preset-env', {targets: {node: 'current'}}]],
+};
```
Minor style note: adding empty at the end of the file.

catheraaine replied:
If you use VS code you can set it to do this on save

catheraaine replied:
This one

![Screen Shot 2021-04-26 at 4 35 12 PM](https://user-images.githubusercontent.com/7475148/116147423-b1f35780-a6ad-11eb-8173-b619c71d1a8e.png)

Also, you should set this to two spaces
![Screen Shot 2021-04-26 at 4 34 52 PM](https://user-images.githubusercontent.com/7475148/116147451-bc155600-a6ad-11eb-9152-348c4da71266.png)

FlashGooden replied:
Noted will do.

jordanleven commented on Commands/Play.js:
```diff
@@ -1,7 +1,21 @@
+const mp3Path = path.resolve(
+   "/Users/flashgooden/Desktop/Sparkbox Development/apprentice-repo/discord-audio-bot/assets/CatSound.mp3"
+);
+const streamPath = "http://www.sample-videos.com/audio/mp3/wave.mp3";
```
One thing to consider: static constants can be noted in UPPER_SNAKE_CASE to indicate that they're static and nonvariate [Google style guide](https://google.github.io/styleguide/javascriptguide.xml#Constants).

jordanleven commented on Commands/Play.js:
```diff
@@ -4,15 +4,20 @@ const mp3Path = path.resolve(
       });
+};
+
+const userInVoiceChannel = (message)=> message.member.voice.channel ? true : false
```
This looks good! Consider using the bangbang (`!!`) operator instead of using a ternary.

FlashGooden replied:
I changed this and love this tip J. that call was returning an object that is only present if in a voice chat, so as long as the object is Truthy this works.

jordanleven commented on Commands/Play.js:
```diff
@@ -1,23 +1,75 @@
+// const mp3Path = path.resolve( "/Users/flashgooden/Desktop/Sparkbox Development/apprentice-repo/discord-audio-bot/assets/CatSound.mp3");
+
+const playMp3 = async (roomConnection, arg) => {
+   arg === "cat"
```
Consider using a switch statement for looking for matches with `arg` but declaring the sound variable at the top and using a switch statement to match the name will allow you to redeclare the value with a match and then run the `play` method instead of redeclaring it.

FlashGooden replied:
Changed this as well.... As much as I love the ternary operator, I like performance a whole lot better.

jordanleven commented on Commands/Play.js:
```diff
@@ -1,23 +1,75 @@
-const userInVoiceChannel = (message)=> message.member.voice.channel ? true : false
-const roomConnection = (message) => message.member.voice.channel.join()
+const userInVoiceChannel = (message) =>
+   message.member.voice.channel ? true : false;
```
Consider using `!!`

jordanleven commented on Commands/Play.spec.js:
```diff
@@ -0,0 +1,30 @@
+            member: { voice: { channel: false } },
+         };
+      });
+      mockArg.mockReturnValue("cat");
```
Consider returning this value inside of the `markArg`

jordanleven commented on Commands/Play.spec.js:
```diff
@@ -0,0 +1,30 @@
+
+describe("Tests the play function", () => {
+   it("will error if you are not in a voice channel", async () => {
+      const mockArg = jest.fn();
```
Do you need this to be a mock function or would an anon function work?

FlashGooden replied:
Anon would def work. Think I was a lil nervous about what to use from Jest initially with mocking, but it's coming together for me now.

jordanleven commented on Commands/Play.spec.js:
```diff
@@ -0,0 +1,30 @@
+import play from "./Play";
+
+describe("Tests the play function", () => {
+   it("will error if you are not in a voice channel", async () => {
```
Does this need to be async?

jordanleven commented on Commands/Play.spec.js:
```diff
@@ -0,0 +1,30 @@
+      expect(mockMessage).toHaveBeenCalledWith("You need to join a voice channel first!")
+   });
+
+   it("will join channel to play mp3 if command is available and user is in room", async () => {
```
You might not need async here.

jordanleven commented on Commands/Play.spec.js:
```diff
@@ -0,0 +1,30 @@
+
+   it("will join channel to play mp3 if command is available and user is in room", async () => {
+      let joinedChannel = 0;
+      const mockMessage = jest.fn();
```
👏🏻  Awesome use of mock functions here!

jordanleven commented on Commands/Help.js:
```diff
@@ -1,7 +1,7 @@
+  message.channel.send('Thank you for using the SoundWave Bot. \n The current features that you can use are \n "!play": to play any sound in the directory \n "!list": to list all songs in the directory \n "!help": well that is what got you here, so type again to see this list of commands')
 }
 
 export default help 
```
Nitpick: this wasn't part of this PR, but I'd consider putting empty lines at the end of your files.

jordanleven commented on Commands/List.js:
```diff
@@ -1,7 +1,7 @@
+  message.channel.send('the current sounds in the directory are \n "cat" \n "fight" \n "replay"')
 }
 
 export default list
```
Nitpick: this wasn't part of this PR, but I'd consider putting empty lines at the end of your files.

jordanleven reviewed (commented):
Hey @Gamerman2001  - looking really awesome! I haven't used Discord before so it was really awesome seeing the capabilities through your PR. Since we went over this in person, I didn't comment on a lot of the things I thought were really great (like the commit messages and the readability on some of the smaller functions) but this is a really great start for your app!

I finished the review after we signed off and left some comments throughout the PR of things that you might want to consider. Here are some of the main themes:

1. Try to keep functions concise and prescriptive. This will help in two areas: isolating your application code from the framework that communicates with Discord (making tests easier) and making the code more readable. I'm, personally, a fan of longer-named functions that tell you what it's returning and what it intakes, and adopting that kind of pattern here will improve readability.
2. Semi-related, try and reduce the surface area between the discord API and your application code. Consider something like a wrapper implementation.

catheraaine commented on Commands/Play.js:
```diff
@@ -1,7 +1,78 @@
+      ? roomConnection
+           .play(await fireBaseJazz(), { volume: 1 })
+           .on("finish", () => {
+              console.log("Finished playing!");
```
you could even make a little function for this part, like 

```js

function playCompletedMessage  {
 console.log("Finished playing!")
}

playCompletedMessage();
```

FlashGooden committed:
refactor: refactor playMp3
- refactored playMp3 to use a switch statement instead of teranaries

FlashGooden committed:
style: styling fix

FlashGooden committed:
chore: added linting scripts and setup

FlashGooden committed:
style: styling adjustments

FlashGooden committed:
test: updated play test
- updated description of the play function
- removed unecessary async await on test

FlashGooden committed:
style: styling fixed
